lib/logstash/filters/date.rb

```diff
@@ -20,256 +25,244 @@
+      config :target, :validate => :string, :default => "@timestamp"
+
+      # Append values to the `tags` field when there has been no
+      # successful match
+      config :tag_on_failure, :validate => :array, :default => ["_dateparsefailure"]
+
+      # LOGSTASH-34
+      DATEPATTERNS = %w{ y d H m s S }
+
+      def initialize(config = {})
+        super
+        @parsers = Hash.new { |h,k| h[k] = [] }
+      end # def initialize
+
+      def register
+        if @match.length < 2
+          error_msg = "The match setting should contains first a field name and at least one date format, current value is #{@match}"
+          raise LogStash::ConfigurationError, i18n(error_msg)
+        end
+
+        replace_underscore(@locale) if @locale && @locale.include?("_")
+        locale = @locale ? Locale.forLanguageTag(@locale) : nil
+
+        @sprintf_timezone = @timezone && !@timezone.index("%{").nil?
+
+        field    = @config["match"][0]
+        patterns = @config["match"][1..-1]
+        setup_parser(field, locale, patterns)
       end
-      locale = java.util.Locale.forLanguageTag(@locale)
-    end
 
-    @sprintf_timezone = @timezone && !@timezone.index("%{").nil?
+      def filter(event)
+        @logger.debug? && @logger.debug("Date filter: received event", :type => event["type"])
 
-    setupMatcher(@config["match"].shift, locale, @config["match"] )
-  end
+        @parsers.each do |field, parsers|
+          @logger.debug? && @logger.debug("Date filter looking for field",
+                                          :type => event["type"], :field => field)
+          next unless event.include?(field)
+          value = event[field]
 
-  def setupMatcher(field, locale, value)
-    value.each do |format|
-      parsers = []
-      case format
-        when "ISO8601"
-          iso_parser = org.joda.time.format.ISODateTimeFormat.dateTimeParser
-          if @timezone && !@sprintf_timezone
-            iso_parser = iso_parser.withZone(org.joda.time.DateTimeZone.forID(@timezone))
-          else
-            iso_parser = iso_parser.withOffsetParsed
-          end
-          parsers << lambda { |date| iso_parser.parseMillis(date) }
-          #Fall back solution of almost ISO8601 date-time
-          almostISOparsers = [
-            org.joda.time.format.DateTimeFormat.forPattern("yyyy-MM-dd HH:mm:ss.SSSZ").getParser(),
-            org.joda.time.format.DateTimeFormat.forPattern("yyyy-MM-dd HH:mm:ss.SSS").getParser()
-          ].to_java(org.joda.time.format.DateTimeParser)
-          joda_parser = org.joda.time.format.DateTimeFormatterBuilder.new.append( nil, almostISOparsers ).toFormatter()
-          if @timezone && !@sprintf_timezone
-            joda_parser = joda_parser.withZone(org.joda.time.DateTimeZone.forID(@timezone))
-          else
-            joda_parser = joda_parser.withOffsetParsed
+          begin
+            epochmillis = -1
+            success = false
+            last_exception = RuntimeError.new "Unknown"
+
```

Comment:
Consider using a holder instance here. e.g.
```ruby
class ParserResult
  attr_reader :event, :field, :date
  attr_accessor :epochmillis, :success, :exception

  def initialize(event, field, exception)
    @event, @field, @exception = event, field, exception
    @epochmillis, @success = -1, false
    @date = event[field]
  end
end

def build_unix_parser
  lambda do |result|
    result.success = /^\d+(?:\.\d+)?$/ =~ result.date || result.date.is_a?(Numeric)
    if result.success
      result.epochmillis = (date.to_f * 1000).to_i
    else
      result.exception = RuntimeError.new("Invalid UNIX epoch value '#{date}'")
    end
  end
end
```
Then the different parsers (lambdas and objects) can have the same interface #call(result).
You will not need to test for format, you can build the parsers array with parsers only not a hash.
The parsers can internalise their rescue and set exception, success and epochmillis as necessary.

then you have...
```ruby
...
result = ParserResult.new(event, field, RuntimeError.new("No parsers "))
parsers.each do |parser|
  parser.call(result)
  break if result.success
end
raise result.exception if !success
event[@target] = LogStash::Timestamp.at(result.epochmillis / 1000, (result.epochmillis % 1000) * 1000)
...
```
WDYT?